### PR TITLE
Add status-go tests

### LINKED ISSUE #13335: Create node bindings of status-go that can be run in tests
As a developer
I want to have node bindings of status-go that can be run in clojure tests
So that I can run tests validating clojure & status-go code together

### Description

In order to increase the test surface and make sure status-react works with a given version of status-go, we want to have some node bindings and be able to run them in clojure tests.
A very basic test would be to pull the right version of status-go when running `make test` and hit any of the c bindings function call.

Some work has already been done in the past https://github.com/status-im/status-nodejs , so this is technically possible.

### PR DESCRIPTION
**Pull request by cammellos**
Fixes: #13335 
status: ready

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -286,6 +290,8 @@ lint-fix: ##@test Run code style checks and fix issues
 
 test: export TARGET := clojure
 test: ##@test Run tests once in NodeJS
+	nix/scripts/build.sh targets.status-go.library
```
Comment (jakubgs) [inline Makefile]: Why do this if you can just make this target depend on the library target?
```suggestion
test: status-go-library ##@test Run tests once in NodeJS
	nix/scripts/build.sh targets.status-go.library
```

File: nix/status-go/library/default.nix [hunk 2]
```diff
@@ -0,0 +1,87 @@
+{ lib, stdenv, utils, callPackage, buildGo116Package
+, go, xcodeWrapper
+# object with source attributes
+, meta, source
+, goBuildFlags ? [ ]
```
Comment (jakubgs) [inline nix/status-go/library/default.nix]: You are receiving and passing `goBuildFlags` but not using them, is that intentional or did you remove too much?

File: nix/status-go/library/default.nix [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  # Ensure XCode is present for iOS build, instead of failing at the end of the build
+  preConfigure = optionalString isDarwin utils.enforceXCodeAvailable;
+
+  # Build mobile libraries
+  preBuild = let
+    NIX_GOWORKDIR = "$NIX_BUILD_TOP/go-build";
+  in ''
+    mkdir ${NIX_GOWORKDIR}
+
+    export GO111MODULE=off
```
Comment (jakubgs) [inline nix/status-go/library/default.nix]: I don't think we need this setting here. Actually, I don't think we need this whole `preBuild` phase.

File: nix/status-go/library/default.nix [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+  # Build the Go library
+  buildPhase = let
+    ldFlags = [ "-extldflags=-Wl,--allow-multiple-definition" ]
+      ++ lib.optionals (secretsFile != "") ["-X node.OpenseaKeyFromEnv=$OPENSEA_API_KEY"]
+      ++ goBuildLdFlags;
+    CGO_LDFLAGS = concatStringsSep " " ldFlags;
+  in ''
+    runHook preBuild
+    runHook renameImports
```
Comment (jakubgs) [inline nix/status-go/library/default.nix]: What is `renameImports`? I don't see it anywhere.

File: nix/status-go/library/default.nix [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    ls $out/
+
+    rm -rf $NIX_GOWORKDIR
+    rm $out/main.go
+
+    runHook postBuild
+  '';
+
+  # replace hardcoded paths to go package in /nix/store, otherwise Nix will fail the build
+  fixupPhase = ''
```
Comment (jakubgs) [inline nix/status-go/library/default.nix]: It doesn't seem like `fixupPhase` is necessary, as there is no references to Go path in the library:
```
post-installation fixup
find: 'remove-references-to': No such file or directory
```
We can check that manually:
```
 > strings /nix/store/d4ys954p76y5cxy7nqii4g4rifl8icd4-status-go-develop-a244d77/libstatus.a | grep '/nix/store/[a-z0-9]+-go'

```

File: nix/status-go/library/default.nix [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  in ''
+    runHook preBuild
+    runHook renameImports
+
+    mkdir -p $out
+    cd ${source.src} && go run ${source.src}/cmd/library/*.go > $out/main.go
+
+    go build -buildmode=c-archive -o $out/libstatus.a $out/main.go
+
+    ls $out/
```
Comment (jakubgs) [inline nix/status-go/library/default.nix]: Why is this here?

File: nix/status-go/library/default.nix [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+    rm -rf $NIX_GOWORKDIR
+    rm $out/main.go
```
Comment (jakubgs) [inline nix/status-go/library/default.nix]: Completely unnecessary. Nix builds in a sandbox, once build is over the sandbox is gone.

File: nix/status-go/library/default.nix [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    ldFlags = [ "-extldflags=-Wl,--allow-multiple-definition" ]
+      ++ lib.optionals (secretsFile != "") ["-X node.OpenseaKeyFromEnv=$OPENSEA_API_KEY"]
+      ++ goBuildLdFlags;
+    CGO_LDFLAGS = concatStringsSep " " ldFlags;
+  in ''
+    runHook preBuild
+    runHook renameImports
+
+    mkdir -p $out
+    cd ${source.src} && go run ${source.src}/cmd/library/*.go > $out/main.go
```
Comment (jakubgs) [inline nix/status-go/library/default.nix]: Why are you putting `main.go` into `$out`? We don't need it there. If it's part of the build it should be under `/build`.

File: nix/status-go/library/default.nix [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    export GO111MODULE=off
+    export NIX_GOWORKDIR=${NIX_GOWORKDIR}
+  '';
+
+  # Build the Go library
+  buildPhase = let
+    ldFlags = [ "-extldflags=-Wl,--allow-multiple-definition" ]
+      ++ lib.optionals (secretsFile != "") ["-X node.OpenseaKeyFromEnv=$OPENSEA_API_KEY"]
+      ++ goBuildLdFlags;
+    CGO_LDFLAGS = concatStringsSep " " ldFlags;
```
Comment (jakubgs) [inline nix/status-go/library/default.nix]: This has no effect. Variables set in a `let` block of a phase(or anything really) are available in the following block to be used, but __are not an env variables in the build__. To have that take effect you have to either:

* Define `CGO_LDFLAGS` in the main argument set for `buildGo116Package`.
* Use `export CGO_LDFLAGS=${CGO_LDFLAGS}` in the `buildPhase` or some other phase.

File: nix/status-go/library/default.nix [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  version = "${source.cleanVersion}-${source.shortRev}";
+
+  inherit meta;
+  inherit (source) src goPackagePath;
+
+  nativeBuildInputs = [ ]
+    ++ optional isDarwin xcodeWrapper;
+
+  # Fixes Cgo related build failures (see https://github.com/NixOS/nixpkgs/issues/25959 )
+  hardeningDisable = [ "fortify" ];
```
Comment (jakubgs) [inline nix/status-go/library/default.nix]: Pretty sure this is also unnecessary.

File: nix/status-go/library/default.nix [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+let
+  inherit (stdenv) isDarwin;
+  inherit (lib)
+    concatStrings concatStringsSep concatMapStrings optionalString
+    getAttr attrValues makeBinPath optional;
+
+  removeReferences = [ go ];
+  removeExpr = refs: ''remove-references-to ${concatMapStrings (ref: " -t ${ref}") refs}'';
+
+in buildGo116Package {
```
Comment (jakubgs) [inline nix/status-go/library/default.nix]: We should probably switch to using `buildGo117Package`.

File: nix/status-go/library/default.nix [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+{ secretsFile ? "" }:
+
+let
+  inherit (stdenv) isDarwin;
+  inherit (lib)
+    concatStrings concatStringsSep concatMapStrings optionalString
+    getAttr attrValues makeBinPath optional;
+
+  removeReferences = [ go ];
+  removeExpr = refs: ''remove-references-to ${concatMapStrings (ref: " -t ${ref}") refs}'';
```
Comment (jakubgs) [inline nix/status-go/library/default.nix]: If we don't run `fixupPhase` we don't need this either.

**Review by jakubgs — CHANGES_REQUESTED** [review]
Good stuff man.

File: nix/status-go/library/default.nix (hunk 9 above)
Comment (jakubgs) [reply nix/status-go/library/default.nix]: A proper way would look something like this:
```nix
{
  buildPhase = let
    ldFlags = [ "-extldflags=-Wl,--allow-multiple-definition" ]
      ++ lib.optionals (secretsFile != "") ["-X node.OpenseaKeyFromEnv=$OPENSEA_API_KEY"]
      ++ goBuildLdFlags;
  in ''
    runHook preBuild
    go build \
      -buildmode=c-archive \
      -ldflags="${concatStringsSep " " ldFlags}" \
      -o libstatus.a main.go
    runHook postBuild
  '';
}
```

File: nix/status-go/library/default.nix (hunk 8 above)
Comment (jakubgs) [reply nix/status-go/library/default.nix]: Athough the more proper way is to use `NIX_BUILD_TOP` variable, as in :` > $NIX_BUILD_TOP/main.go`

Comment (jakubgs) [reply nix/status-go/library/default.nix]: Also, this part should probably be moved to a `preBuild` hook:
```nix
{
  preBuild = ''
    pushd go/src/$goPackagePath
    go run cmd/library/*.go > $NIX_BUILD_TOP/main.go
    popd
  '';
}
```

File: nix/status-go/library/default.nix (hunk 2 above)
Comment (cammellos) [reply nix/status-go/library/default.nix]: Not using them, removing

**Comment by jakubgs** [comment]
I verified on a CI host that `make status-go-library` builds fine on MacOS without `nativeBuildInputs` and `preConfigure`.

**Review by jakubgs — APPROVED** [review]
Cool stuff.

**Pull request merged**